spec.html

```diff
@@ -32706,61 +32677,42 @@ <h1>The Array Constructor</h1>
-          1. If Type(_len_) is not Number, then
-            1. Perform ! CreateDataPropertyOrThrow(_array_, *"0"*, _len_).
-            1. Let _intLen_ be 1.
-          1. Else,
-            1. Let _intLen_ be ToUint32(_len_).
-            1. If _intLen_ &ne; _len_, throw a *RangeError* exception.
-          1. Perform ! Set(_array_, *"length"*, _intLen_, *true*).
-          1. Return _array_.
-        </emu-alg>
-      </emu-clause>
-
-      <emu-clause id="sec-array-items">
-        <h1>Array ( ..._items_ )</h1>
-        <p>This description applies if and only if the Array constructor is called with at least two arguments.</p>
+      <emu-clause id="sec-array" oldids="sec-array-constructor-array,sec-array-len,sec-array-items">
+        <h1>Array ( ..._values_ )</h1>
         <p>When the `Array` function is called, the following steps are taken:</p>
         <emu-alg>
-          1. Let _numberOfArgs_ be the number of arguments passed to this function call.
-          1. Assert: _numberOfArgs_ &ge; 2.
           1. If NewTarget is *undefined*, let _newTarget_ be the active function object; else let _newTarget_ be NewTarget.
           1. Let _proto_ be ? GetPrototypeFromConstructor(_newTarget_, *"%Array.prototype%"*).
-          1. Let _array_ be ? ArrayCreate(_numberOfArgs_, _proto_).
-          1. Let _k_ be 0.
-          1. Repeat, while _k_ &lt; _numberOfArgs_,
-            1. Let _Pk_ be ! ToString(_k_).
-            1. Let _itemK_ be _items_[_k_].
-            1. Perform ! CreateDataPropertyOrThrow(_array_, _Pk_, _itemK_).
-            1. Set _k_ to _k_ + 1.
-          1. Assert: The value of _array_'s *"length"* property is _numberOfArgs_.
-          1. Return _array_.
+          1. Let _numberOfArgs_ be the number of elements in _values_.
+          1. If _numberOfArgs_ = 0, then
+            1. Return ! ArrayCreate(0, _proto_).
+          1. Else if _numberOfArgs_ = 1, then
+            1. Let _len_ be _values_[0].
+            1. Let _array_ be ! ArrayCreate(0, _proto_).
+            1. If Type(_len_) is not Number, then
+              1. Perform ! CreateDataPropertyOrThrow(_array_, *"0"*, _len_).
+              1. Let _intLen_ be 1.
+            1. Else,
+              1. Let _intLen_ be ToUint32(_len_).
+              1. If _intLen_ &ne; _len_, throw a *RangeError* exception.
+            1. Perform ! Set(_array_, *"length"*, _intLen_, *true*).
+            1. Return _array_.
+          1. Else,
+            1. Assert: _numberOfArgs_ &ge; 2.
+            1. Let _array_ be ? ArrayCreate(_numberOfArgs_, _proto_).
+            1. Let _k_ be 0.
+            1. Repeat, while _k_ &lt; _numberOfArgs_,
+              1. Let _Pk_ be ! ToString(_k_).
+              1. Let _itemK_ be _values_[_k_].
+              1. Perform ! CreateDataPropertyOrThrow(_array_, _Pk_, _itemK_).
+              1. Set _k_ to _k_ + 1.
+            1. Assert: The value of _array_'s *"length"* property is _numberOfArgs_.
+            1. Return _array_.
```

Comment:
Well, we can't hoist the Assert, because it's not generally true in the case of `_numberOfArgs_ = 1`.

But we could hoist the Return. I think I looked at that and decided against it: since each if-arm creates an array, it seemed more self-contained for the if-arm to also return it. But it's not a big preference.

Reply:
aha, true enough. i guess i have a light preference for hoisting all shared steps, but it's fine as-is too.

Reply:
Thanks

Reply:
I prefer the current form; moving the `return`s out of the branches makes it harder for me to follow the flow.